src/pint_pal/noise_utils.py

```diff
@@ -24,17 +31,120 @@
 from enterprise.signals import deterministic_signals
 from enterprise import constants as const
 
+from enterprise_extensions.sampler import group_from_params, get_parameter_groups
 from enterprise_extensions import model_utils
 from enterprise_extensions import deterministic
 from enterprise_extensions.timing import timing_block
-#from enterprise_extensions.blocks import (white_noise_block, red_noise_block)
+try:
+    import xarray as xr
+    import jax
+    from jax import numpy as jnp
+    import numpyro
+    from numpyro.infer import log_likelihood
+    from numpyro import distributions as dist
+    from numpyro import infer
+    import discovery as ds
+    from discovery import prior as ds_prior
+    from discovery.prior import (makelogtransform_uniform,
+                                 makelogprior_uniform,
+                                 sample_uniform)
+    from discovery.gibbs import setup_single_psr_hmc_gibbs
+
+except ImportError:
+    log.error("Please install the latest version of discovery, numpyro, and/or jax")
+    ValueError("Please install the latest version of discovery, numpyro, and/or jax")
```

Comment:
The `ValueError()` line isn't doing anything useful here, it just defines an exception object and does nothing with it. If you actually wanted to raise an exception you'd have to write `raise ValueError(...)`. But I don't recommend doing that, since it would make everything break if you imported `noise_utils` without these packages installed.

I think what you want to do is make as many of these as you can into explicit dependencies, and for the ones that you can't (Discovery), import them as close as possible to where they're actually used, for example inside `setup_discovery_noise()` or inside `model_noise()`, in the case where `likelihood="discovery"`. That way most things can run even without the package installed, and only the things that really need it will break if it is missing.

Reply:
do we want jax and numpyro to be explicit dependencies of pintpal? they are necessary to run ‘model_noise’ with the ‘discovery’ likelihood.

Reply:
I don't think it would hurt much to do so, but since we can't make discovery an explicit dependency for now, it might make more sense to import them only where they are needed (inside `model_noise`, and only in the case where `likelihood="discovery"`).

Whatever we do, I'd like to make sure of two things: (1) anything that is imported unconditionally should be an explicit dependency, and (2) it should still be possible to use PINT Pal with the Enterprise likelihood without getting any errors, even if Discovery is not installed.

Reply:
ok for now, i guess let’s leave jax and numpyro out of the requirements since you can’t use them without discovery anyways. i can confirm at this point that everything imported unconditionally is listed in the dependencies and that you can still use enterprise/ptmcmcsampler without discovery.